### PR TITLE
[test][msan] Precommit tests for vararg improvements

### PR DESCRIPTION
**Pull request by vitalybuka**

### CODE REVIEW

File: compiler-rt/test/msan/vararg_shadow.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  fprintf(stderr, "\nTLS at line %d: ", line);
+  for (int i = 0; i < kMsanParamTlsSize;) {
+    fprintf(stderr, "\n");
+    for (int j = 0; j < 16; ++i, ++j)
+      fprintf(stderr, "%02x ", tmp[i]);
+  }
+
+  fprintf(stderr, "\n");
+}
+#endif
```
Comment (thurstond) [inline compiler-rt/test/msan/vararg_shadow.cpp:84]: Please add `// DEBUG_TEST`

**Review by kstoimenov — COMMENTED** [review]
Why do we need separate ll files for ARM and X86? Is there a way to combine those?

File: compiler-rt/test/msan/vararg_shadow.cpp [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+template <class T>
+__attribute__((noinline)) void print_shadow(va_list &args, int n,
+                                            const char *function) {
+  for (int i = 0; i < n; i++) {
+    // 1-based to make it different from clean shadow.
+    fprintf(stderr, "\nArgShadow fn:%s n:%d i:%02x ", function, n, i + 1);
+    T arg_int = va_arg(args, T);
+    if (__msan_test_shadow(&arg_int, sizeof(arg_int)))
+      fprintf(stderr, "fale[clean] %02x", i + 1);
```
Comment (thurstond) [inline compiler-rt/test/msan/vararg_shadow.cpp:129]: Is "fale" a typo?

File: compiler-rt/test/msan/vararg_shadow.cpp [hunk 3]
```diff
@@ -0,0 +1,256 @@
+// Check that shadow of retrived value from va_list matches the shadow of passed value.
```
Comment (thurstond) [inline compiler-rt/test/msan/vararg_shadow.cpp:1]: Nit: "retrieved"

**Comment by vitalybuka** [comment]
> Why do we need separate ll files for ARM and X86? Is there a way to combine those?

VarArg is ABI specific
IR is different even before sanitizer instrumentation.

File: compiler-rt/test/msan/vararg_shadow.cpp [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+// Nothing works yet.
+// XFAIL: *
+
+#include <sanitizer/msan_interface.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#define DEBUG_TEST 0
```
Comment (thurstond) [inline compiler-rt/test/msan/vararg_shadow.cpp:15]: Maybe guard it with ifdef, to allow overriding it as a compile flag?

**Review by thurstond — APPROVED** [review]

File: compiler-rt/test/msan/vararg_shadow.cpp (hunk 2 above)
Comment (vitalybuka) [reply compiler-rt/test/msan/vararg_shadow.cpp:129]: No, it's type from fake :)

**Pull request merged**